provider/provider.go

```diff
@@ -70,6 +78,50 @@ func NewPinnedProvider(onlyRoots bool, pinning pin.Pinner, fetchConfig fetcher.F
 	}
 }
 
+// NewBufferedProvider returns a KeyChanFunc supplying keys from a given
+// KeyChanFunction, but buffering all the keys in memory first until they are
+// read.  This allows the underlying KeyChanFunc to finish listing pins as
+// soon as possible releasing any resources, locks, at the expense of memory
+// usage.
+func NewBufferedProvider(pinsF KeyChanFunc) KeyChanFunc {
+	return func(ctx context.Context) (<-chan cid.Cid, error) {
+		pins, err := pinsF(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		// Use very efficient queue implementation for storing the cids.
+		queue := new(deque.Deque[cid.Cid])
+		queue.SetBaseCap(1024)
+
+		totalCids := 0
+		for c := range pins {
+			queue.PushBack(c)
+			totalCids++
+		}
```

Comment:
If this section was run in a goroutine, and was writing to the queue, then the cids could be read from the queue without having wait for all the pins to be enqueued first.  Doing this would mean protecting the queue with synchronization. If there were something that looked like an infinite-length channel, that would work. See [`chanqueue`](https://github.com/gammazero/chanqueue?tab=readme-ov-file#chanqueue)

Reply:
I should have seen `chanqueue` before. Let's switch to it.